src/nti/webhooks/subscriptions.py

```diff
@@ -248,11 +374,39 @@ def prune_subscription_when_resolved(event):
                 break
 
 
+@component.adapter(IWebhookDeliveryAttemptFailedEvent)
+def deactivate_subscription_when_all_failed(event):
+    # type: (IWebhookDeliveryAttemptFailedEvent) -> None
+    attempt = event.object # type: WebhookDeliveryAttempt
+    subscription = attempt.__parent__
+    if not _subscription_full(subscription, True):
+        return
+    # XXX: Add logging here.
+    # This is a very simple-minded approach. Something more featured
+    # might involve a ratio of failed attempts? Over some sort of sliding window?
+    # Or examining the time period?
+    # This has to activate all the sub-objects, which could be expensive.
+    # We could make the subscription use BTree Length objects to track
+    # the various states.
+    if all(attempt.failed() for attempt in subscription.values()):
+        manager = subscription.__parent__ # type:PersistentWebhookSubscriptionManager
+        manager.deactivateSubscription(subscription)
```

Comment:
Does this notify that it has been deactivated? Wondering where we might hook into sending a notification to the owner that their subscription is deactivated.

Reply:
It does, yeah, though I admit it's not really obvious. We can introduce something more obvious if we need to.

https://github.com/NextThought/nti.webhooks/blob/9ccda01609052c26b8e6b2a9c61d769fff4e5073/src/nti/webhooks/subscriptions.py#L291-L297